src/generic.jl

```diff
@@ -32,39 +32,58 @@ evaluate(dist::PreMetric, a, b) = dist(a, b)
 Infer the result type of metric `dist` with input type `Ta` and `Tb`, or input
 data `a` and `b`.
 """
-result_type(::PreMetric, ::Type, ::Type) = Float64 # fallback in Distances
+# result_type(::PreMetric, ::Type, ::Type) = Float64 # fallback in Distances
 result_type(f, a::Type, b::Type) = typeof(f(oneunit(a), oneunit(b))) # don't require `PreMetric` subtyping
 
-# Promote Arrays and Numbers to types
 result_type(dist, a, b) = result_type(dist, _eltype(a), _eltype(b))
+# description of approach:
+# (a) for generic iterators, rely on Base.IteratorEltype(a)
 _eltype(a) = __eltype(Base.IteratorEltype(a), a)
-__eltype(::Base.HasEltype, a) = eltype(a)
-__eltype(::Base.EltypeUnknown, a) = typeof(first(a))
+__eltype(::Base.HasEltype, a) = _eltype(eltype(a))
+__eltype(::Base.EltypeUnknown, a) = _eltype(typeof(first(a)))
+# (b) for arrays (of arrays), go down the eltype route, until you hit a non-array eltype and return that
+_eltype(a::AbstractArray) = _eltype(eltype(a))
```

Comment:
Is it really needed to special-case arrays? The generic code for iterators should work AFAICT.

Also, do you think arrays of arrays of arrays will work? If we want to allow that, maybe there should be a test?

Reply:
Yes, without that specialization, you only ask for the `eltype` once, and in the case of, say, vectors of vectors you get some array type. Then, `result_type` tries to construct a `oneunit` of that array type... and fails. This is consistent with the explicit `eltype(eltype(a))` construction in #188.

Reply:
And we do have tests for the array of array case, basically copied from #188.

Reply:
What I mean is that for arrays, `_eltype(a) = __eltype(Base.IteratorEltype(a), a)` should end up calling `__eltype(::Base.HasEltype, a) = _eltype(eltype(a))`, which is exactly the same as this definition. Am I missing something?

Reply:
I see. I removed the special casing now, assuming that the "bottom type" is reached when `eltype(T) === T`. Otherwise, both arrays (of arrays/iterators) and iterators (of arrays/iterators) should go down the route until you hit that condition.